### PR TITLE
[#205] Porting command line parsing infrastructure

### PR DESCRIPTION
**Pull request by lenhat509**
Porting CLI infrastructure from pgagroal to pgmoneta

### CODE REVIEW

**Comment by lenhat509** [comment]
Hi @jesperpedersen @Jubilee101, I completed porting the command line parsing feature from pgagroal. Let me know if there is any concern.

File: src/libpgmoneta/utils.c [hunk 1]
```diff
@@ -3003,6 +3004,128 @@ pgmoneta_is_file_archive(char* file_path)
...
+      }
+      return false;
+   }
+
+   parsed->cmd = &command_table[command_index];
+
+   /* Iterate until find an accepted_arg_count that is equal or greater than the typed command arg_count */
+   arg_count = argc - offset;
+   int j;
+   for (j = 0; j < MISC_LENGTH; j++)
```
Comment (jesperpedersen) [inline src/libpgmoneta/utils.c:3089]: Define `int j` in the `for` loop

Comment (jesperpedersen) [reply src/libpgmoneta/utils.c:3089]: Or move it to the top of the function

File: src/libpgmoneta/utils.c [hunk 2]
```diff
@@ -3003,6 +3004,128 @@ pgmoneta_is_file_archive(char* file_path)
...
+/* Parser for pgmoneta-cli amd pgmoneta-admin commands */
+bool
+parse_command(int argc,
+              char** argv,
+              int offset,
+              struct pgmoneta_parsed_command* parsed,
+              const struct pgmoneta_command command_table[],
+              size_t command_count)
+{
+#define EMPTY_STR(_s) (_s[0] == 0)
```
Comment (jesperpedersen) [inline src/libpgmoneta/utils.c:3016]: Hmm, isn't

```
if (s != NULL && strlen(s) > 0)
```

better ?

File: src/libpgmoneta/utils.c [hunk 3]
```diff
@@ -53,6 +53,7 @@
 #include <sys/time.h>
 #include <sys/stat.h>
 #include <sys/types.h>
+#include <err.h>
```
Comment (jesperpedersen) [inline src/libpgmoneta/utils.c:56]: Sorting

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

File: src/libpgmoneta/utils.c (hunk 1 above)
Comment (lenhat509) [reply src/libpgmoneta/utils.c:3089]: Got it. I'll move j to the top since it is used later.

File: src/libpgmoneta/utils.c (hunk 3 above)
Comment (lenhat509) [reply src/libpgmoneta/utils.c:56]: I'll fix that. Thanks.

File: src/libpgmoneta/utils.c (hunk 2 above)
Comment (lenhat509) [reply src/libpgmoneta/utils.c:3016]: I think it makes the code a bit cleaner if we use macros. Subcommand is usually empty and we do the check frequently.

**Comment by jesperpedersen** [comment]
You need to update `contrib/shell_comp`

File: src/libpgmoneta/utils.c (hunk 2 above)
Comment (jesperpedersen) [inline src/libpgmoneta/utils.c:3016]: I think you should move this to `pgmoneta.h`

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by lenhat509** [comment]
> You need to update `contrib/shell_comp`

Sorry about missing that. I'll work on it.

File: src/libpgmoneta/utils.c (hunk 2 above)
Comment (lenhat509) [reply src/libpgmoneta/utils.c:3016]: Just the macro, not the whole parse_command function, right?

Comment (jesperpedersen) [reply src/libpgmoneta/utils.c:3016]: Just macro

**Comment by lenhat509** [comment]
Hi @jesperpedersen, I updated `contrib/shell_comp`, fixed a bug in `admin.c`, modified `cli.c` to handle failed commands nicely. In `cli.c`, it can display a helper function for the corresponding failed command instead of showing the entire menu. `usage()` will only be executed if the primary command does not exist. As for `admin.c`, the entire menu will be displayed since it is relatively small.

**Comment by jesperpedersen** [comment]
Please, rebase

**Comment by lenhat509** [comment]
I added back version info for deprecated commands. Initially, I removed this info when I ported the code from pgagroal because I did not see version variables like `PGMONETA_MAJOR_VERSION` in pgmoneta. I just realized that I could define them in CMakeList.txt so I updated the code to show the version from which the commands were outdated (0.11).

**Comment by jesperpedersen** [comment]
You need to update the documentation too...

Do a `git grep pgmoneta-cli` to see the files

**Comment by lenhat509** [comment]
Thanks for letting me know. I updated the doc. Let me know if I miss something else.

**Comment by jesperpedersen** [comment]
Please, rebase

**Comment by lenhat509** [comment]
Hi @jesperpedersen , I have rebased it.

**Comment by jesperpedersen** [comment]
Merged.

Thanks for your contribution !

**Pull request merged**